tests/system/src/test/java/test/com/sun/javafx/tk/quantum/SystemMenuBarTest.java

```diff
@@ -0,0 +1,351 @@
+
+        Platform.runLater(() -> {
+            mainMenu.getItems().clear();
+            mainMenu.getItems().add(subMenu);
+            subMenu.getItems().addAll(new MenuItem("new item 1"), new MenuItem("new item 2"));
+            removeMenuLatch.countDown();
+        });
+    }
+
+    @Test
+    public void testJDK8309935() {
+        MenuBar menuBar = new MenuBar();
+        AtomicReference<Throwable> throwableRef = new AtomicReference();
+        Util.runAndWait(() -> {
+            Thread.currentThread().setUncaughtExceptionHandler((t, e) -> {
+                e.printStackTrace();
+                throwableRef.set(e);
+            });
+            menuBar.setUseSystemMenuBar(true);
+            Menu menu1 = new Menu("menu 1");
+            menu1.getItems().add(new MenuItem("item 1"));
+            menu1.getItems().add(new MenuItem("item 2"));
+            menuBar.getMenus().add(menu1);
+            Menu menu2 = new Menu(" menu 2");
+            menu2.getItems().add(new MenuItem("item 1"));
+            menu2.getItems().add(new MenuItem("item 2"));
+            menu2.getItems().add(new SeparatorMenuItem());
+            menuBar.getMenus().add(menu2);
+            Menu test1 = new Menu("test 1");
+            test1.getItems().add(new MenuItem("item 1"));
+            test1.getItems().add(new MenuItem("item 2"));
+            Menu test2 = new Menu("test 2");
+            test2.getItems().add(new MenuItem("item 1"));
+            test2.getItems().add(new MenuItem("item 2"));
+            menu2.addEventFilter(Menu.ON_SHOWING, e -> {
+                menu2.getItems().removeIf(o -> Objects.equals(o.getText(), test1.getText()));
+                menu2.getItems().add(test1);
+                menu2.getItems().removeIf(o -> Objects.equals(o.getText(), test2.getText()));
+                menu2.getItems().add(test2);
+            });
+            BorderPane root = new BorderPane();
+            root.setTop(menuBar);
+            Stage stage = new Stage();
+            stage.setScene(new Scene(root));
+            stage.show();
+        });
+        Util.runAndWait(() -> {
+            menuBar.getMenus().forEach(menu -> {
+                menu.setVisible(false);
+            });
+        });
+        Util.runAndWait(() -> {
+            menuBar.getMenus().forEach(menu -> {
+                menu.setVisible(true);
+            });
+        });
+        Util.runAndWait(() -> {
+            Menu test3 = new Menu("test 3");
+            test3.getItems().add(new MenuItem("item 1"));
+            test3.getItems().add(new MenuItem("item 2"));
+            Menu test4 = new Menu("test 4");
+            test4.getItems().add(new MenuItem("item 1"));
+            test4.getItems().add(new MenuItem("item 2"));
+            menuBar.getMenus().get(1).getItems().addAll(test3, test4);
+        });
+        // Some waiting is necessary. runAndWait twice makes it reliable.
+        Util.runAndWait(() -> {
+        });
+        Util.runAndWait(() -> {
+        });
```

Comment:
I checked the code that `runAndWait` would do when you pass in 0 runnables, and it basically does no waiting at all (less than a microsecond for sure).  If this is really essential (I removed it and the test still passed) then I think a `sleep` might be better.

Reply:
The Util.runAndWait was indeed not a good idea. The problem is not that the test might not pass, the problem is that the test might not fail if we don't wait for things that need to run to complete. But this is better done with a Platform.runLater() which will add a runnable to the queue, so when that is executed, we know that all previously scheduled runnables completed.
I changed the approach ehre.